OneSignalSDK/onesignal/src/main/java/com/onesignal/OutcomeEventsController.java

```diff
@@ -0,0 +1,192 @@
+package com.onesignal;
+
+import android.os.Bundle;
+import android.os.Process;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import org.json.JSONArray;
+
+import java.util.List;
+import java.util.Set;
+
+class OutcomeEventsController {
+
+    private static final String OS_SAVE_OUTCOMES = "OS_SAVE_OUTCOMES";
+
+    static class OutcomeException extends Exception {
+
+        OutcomeException(String message) {
+            super(message);
+        }
+
+    }
+
+    private Set<String> eventsSent = OSUtils.newConcurrentSet();
```

Comment:
looking to the session stuff, this we need eventsSent to live as the session lives right? while the same session is up we should not let the user sent twice the same event. So if the user close and re open the app on the 30 secs of the session in theory i think that wont work?

Reply:
For unique outcomes they should not be sent more than once per notification id. @mikechoch Is making the changes to handle this right now.